SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -146,30 +162,71 @@ void USpatialSender::GainAuthorityThenAddComponent(USpatialActorChannel* Channel
 			// adding the subobject.
 			PendingSubobjectAttachment->PendingAuthorityDelegations.Add(ComponentId);
 			Receiver->PendingEntitySubobjectDelegations.Add(
-				MakeTuple(static_cast<Worker_EntityId_Key>(Channel->GetEntityId()), ComponentId),
+				MakeTuple(static_cast<Worker_EntityId_Key>(EntityId), ComponentId),
 				PendingSubobjectAttachment);
 
-			EntityACL->ComponentWriteAcl.Add(Info->SchemaComponents[Type], AuthoritativeWorkerRequirementSet);
+			NewComponentIds.Add(ComponentId);
 		}
 	});
 
-	FWorkerComponentUpdate Update = EntityACL->CreateEntityAclUpdate();
+	// If this worker is EntityACL authoritative, we can directly update the component IDs to gain authority over.
+	if (StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+	{
+		const FClassInfo& ActorInfo = ClassInfoManager->GetOrCreateClassInfoByClass(Channel->Actor->GetClass());
+		const WorkerAttributeSet WorkerAttribute = { ActorInfo.WorkerType.ToString() };
+		const WorkerRequirementSet AuthoritativeWorkerRequirementSet = { WorkerAttribute };
+
+		EntityAcl* EntityACL = StaticComponentView->GetComponentData<EntityAcl>(Channel->GetEntityId());
+		for (auto& ComponentId : NewComponentIds)
+		{
+			EntityACL->ComponentWriteAcl.Add(ComponentId, AuthoritativeWorkerRequirementSet);
+		}
+
+		FWorkerComponentUpdate Update = EntityACL->CreateEntityAclUpdate();
+		Connection->SendComponentUpdate(Channel->GetEntityId(), &Update);
+	}
+
+	// Update the ComponentPresence component with the new component IDs. If this worker does not have EntityACL
+	// authority, this component is used to inform the enforcer of the component IDs to add to the EntityACL.
+	check(StaticComponentView->HasAuthority(EntityId, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID));
+	ComponentPresence* ComponentPresenceData = StaticComponentView->GetComponentData<ComponentPresence>(EntityId);
+	ComponentPresenceData->AddComponentIds(NewComponentIds);
+	FWorkerComponentUpdate Update = ComponentPresenceData->CreateComponentPresenceUpdate();
 	Connection->SendComponentUpdate(Channel->GetEntityId(), &Update);
 }
 
-void USpatialSender::SendRemoveComponent(Worker_EntityId EntityId, const FClassInfo& Info)
+void USpatialSender::SendRemoveComponentForClassInfo(Worker_EntityId EntityId, const FClassInfo& Info)
 {
+	TArray<Worker_ComponentId> ComponentsToRemove;
 	for (Worker_ComponentId SubobjectComponentId : Info.SchemaComponents)
 	{
 		if (SubobjectComponentId != SpatialConstants::INVALID_COMPONENT_ID)
 		{
 			NetDriver->Connection->SendRemoveComponent(EntityId, SubobjectComponentId);
+			ComponentsToRemove.Add(SubobjectComponentId);
```

Comment:
Yikes - we use `NetDriver->Connection` everywhere in this file, even though we have our own `Connection` member

Reply:
PSYCH that's the wrong type 

![image](https://user-images.githubusercontent.com/9222722/76985795-affe8700-6938-11ea-9fb5-2e7f145ba837.png)

(UNetConnection vs. SpatialWorkerConnection)

Reply:
Oh snap